llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -6397,6 +6397,48 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
       return Constant::getNullValue(ReturnType);
     break;
   }
+  case Intrinsic::ptrmask: {
+    // Fail loudly in case this is ever changed.
+    // TODO: If vector types are supported the logic that checks if the mask is
+    // useless should be updated to use generic constants.
+    assert(!Op0->getType()->isVectorTy() && !Op1->getType()->isVectorTy() &&
+           "These simplifications where written at a time when ptrmask did not "
+           "support vector types and may not work for vectors");
+
+    // NOTE: We can't apply these simplifications based on the value of Op1
+    // because we need to preserve provenance.
+    if (isa<PoisonValue>(Op0) || isa<PoisonValue>(Op1))
+      return PoisonValue::get(Op0->getType());
+
+    if (Q.isUndefValue(Op0))
+      return Constant::getNullValue(Op0->getType());
+
+    if (match(Op0, m_Zero()))
+      return Constant::getNullValue(Op0->getType());
+
+    if (Op1->getType()->getScalarSizeInBits() ==
+        Q.DL.getPointerTypeSizeInBits(Op0->getType())) {
```

Comment:
This is just following the currently LangRef definition (https://llvm.org/docs/LangRef.html#llvm-ptrmask-intrinsic), which refers to the pointer size, not the pointer index size. If you'd like to change the spec to use the pointer index size, then a LangRef patch would be the starting point.

Reply:
Looking at the current langref:
> The result of ptrmask(ptr, mask) is equivalent to getelementptr ptr, (ptrtoint(ptr) & mask) - ptrtoint(ptr).

Since it's defined in terms of GEP it indicates to me that the masking cannot affect more than the index bits. GEP semantics:
>  If the offsets have a different width from the pointer’s index type, they are sign-extended or truncated to the width of the pointer’s index type.

Reply:
@arichardson guess my feeling is for this series (just adding basic support) would prefer to keep checks on the more conservative side. Think its easier to expand than try and do it all at once.

Reply:
Sure, I think keeping it minimal makes sense. I was just wondering if it's correct to use DL.getPointerIndexSizeInBits() here instead? This should be the same behaviour for almost all targets but would make this fold work for the downstream CHERI targets as well once merged.

Reply:
Drop this check now? For the following transforms:

 * The ptrtoint one is safe, but you won't actually be able to trigger it if index type is smaller.
 * The allones/undef ones are safe and can be tested.
 * The knownbits one is safe with a truncate of the known bits. Can be tested.